src/main/java/org/spongepowered/api/resource/ResourcePath.java

```diff
@@ -0,0 +1,246 @@
+     * @return True if this is a child to other
+     */
+    boolean startsWith(ResourcePath parent);
+
+    /**
+     * Gets the file name portion of this {@link ResourcePath}.
+     *
+     * @return The file name of the path
+     * @see #getBaseName()
+     * @see #getExtension()
+     */
+    String getFileName();
+
+    /**
+     * Gets the base name of this path without the extension
+     *
+     * @return The base name
+     * @see #getFileName()
+     */
+    String getBaseName();
+
+    /**
+     * Gets the extension of this path, which is the portion which is after the
+     * last dot if it has one.
+     *
+     * <p>If the path does not have an extension, {@link Optional#empty() empty}
+     * is returned.</p>
+     *
+     * @return The extension of empty if not present
+     * @see #hasExtension(String...)
+     */
+    Optional<String> getExtension();
+
+    /**
+     * Tests if this path has one of the given extensions.
+     *
+     * @param ext The extensions to test
+     * @return True if the extension matches one of the ones provided
+     * @see #getExtension()
+     */
+    boolean hasExtension(String... ext);
+
+    /**
+     * Resolves a new {@link ResourcePath} with this one as the parent.
+     *
+     * @param name The file name of the new path
+     * @return A new path
+     */
+    ResourcePath resolve(String name);
+
+    /**
+     * Resolves a new {@link ResourcePath} with the parent of this one as the
+     * parent.
+     *
+     * @param name The file name of the new path
+     * @return A new path
+     */
+    ResourcePath resolveSibling(String name);
+
+    /**
+     * Provides a String representation of this object consisting of
+     * {@code namespace:path}. This is easily reversible via
+     * {@link #parse(String)}.
+     *
+     * @return A string reprsentation
+     */
+    @Override
+    String toString();
+
+    /**
+     * {@inheritDoc}
+     *
+     * @implNote Implementation should override this for Map usage.
+     */
+    @Override
+    boolean equals(Object other);
+
+    /**
+     * {@inheritDoc}
+     *
+     * @implNote Implementation should override this for Map usage.
+     */
+    @Override
+    int hashCode();
+
+    /**
+     * Represents a builder to create {@link ResourcePath} instances.
+     */
+    interface Builder extends ResettableBuilder<ResourcePath, Builder> {
+
+        /**
+         * Sets the namespace of the {@link ResourcePath}.
+         *
+         * @param namespace The namespace
+         * @return This builder
+         * @throws IllegalArgumentException if the namespace contains illegal characters
+         */
+        Builder namespace(String namespace) throws IllegalArgumentException;
+
+        /**
+         * Sets the namespace of the {@link ResourcePath} using the given
+         * plugin's id.
+         *
+         * @param plugin The owning plugin
+         * @return This builder
+         */
+        Builder namespace(PluginContainer plugin);
```

Comment:
Accept object as well for plugin instance?

Reply:
I was keeping consistent with `CatalogKey.Builder`

Reply:
We don't do Object plugin parameters anymore.